lib/gosu/compat.rb

```diff
@@ -0,0 +1,187 @@
+# This whole file is about backward compatibility
+
+# First - define some methods to tell the world about their old code
+module Gosu
+  module Deprecate
+    WARNING_STACKTRACE_LINES = 1
+
+    # Basically copy&paste from Gem::Deprecate but remove the date-part.
+    def self.deprecate(klass, name, repl)
+      klass.class_eval {
+        old = "_deprecated_#{name}"
+        alias_method old, name
+        define_method name do |*args, &block|
+          Gosu::Deprecate.deprecation_message(self, name, repl)
+          send old, *args, &block
+        end
+      }
+    end
+    def deprecate name, repl
+      Gosu::Deprecate.deprecate(self, name, repl)
+    end
+
+    def self.deprecate_const(klass, name, repl)
+      klass.send(:remove_const, name) if klass.const_defined?(name)
+
+      @@_deprecated_constants ||= {}
+      @@_deprecated_constants[name] = repl
+    end
+    def deprecate_const(name, repl)
+      Gosu::Deprecate.deprecate_const(self, name, repl)
+    end
+
+    # The advantage of this approach is that the old Names won't show up in
+    # any Codecompletion (like IRB) or *.constants, so you have to know the
+    # exact old name to acutally use the old const.
+    # However define the value after the first call, to show the warning only once.
+    def const_missing(const_name)
+      if @@_deprecated_constants && repl = @@_deprecated_constants[const_name]
+        deprecation_message(const_name, repl)
+        const_set const_name, const_get(repl)
+        const_get const_name
+      else
+        super
+      end
+    end
+
+    def self.deprecation_message(klass, name, repl)
+      target = klass.kind_of?(Module) ? "#{klass}::" : "#{klass.class}#"
+      msg = [ "DEPRECATION WARNING: #{target}#{name} is deprecated",
+        repl == :none ? " with no replacement." : "; use #{repl} instead.",
+        "\n#{target}#{name} called from #{Gosu::Deprecate.filtered_caller.join("\n")}",
+      ]
+      warn "#{msg.join}."
+    end
+    def deprecation_message(name, repl)
+      Gosu::Deprecate.deprecation_message(self, name, repl)
+    end
+
+    # Just show a "unified" message without the "magic" of the other two methods.
+    def self.message(msg)
+      msg = [ "DEPRECATION WARNING: #{msg}",
+        "\n#Called from #{caller[1, WARNING_STACKTRACE_LINES]}",
+      ]
+      warn "#{msg.join}."
+    end
+
+    # this method removes the noise of the caller due to the deprecation methods
+    def self.filtered_caller
+      caller.delete_if { |trace_line| trace_line =~ /(deprecat|const_missing)/ }[0, WARNING_STACKTRACE_LINES]
+    end
+  end
+end
+
+# No need to pass a Window to Image, Sample and Song
+class Gosu::Image
+  alias initialize_without_window initialize
+
+  def initialize(*args)
+    if args[0].is_a? Gosu::Window then
+      Gosu::Deprecate.message("Passing a Window to Image#initialize has been deprecated in Gosu 0.9.")
```

Comment:
This is a little misleading because not only the `window` argument has changed, but some arguments have moved into a hash as well. Not sure what to do about this - do you think a link to the rdoc would be too much?